src/cudadev/DataFormats/SoAStore.h

```diff
@@ -0,0 +1,569 @@
+    size_t offset=0;                                                                                                                \
+    _ITERATE_ON_ALL(_DECLARE_SOA_DUMP_INFO, ~, __VA_ARGS__)                                                                         \
+    std::cout << "Final offset = " << offset << " computeDataSize(...): " << computeDataSize(nElements, byteAlignment) << std::endl;\
+    std::cout << std::endl;                                                                                                         \
+  }                                                                                                                                 \
+  /* Helper function used by caller to externally allocate the storage */                                                           \
+  static size_t computeDataSize(size_t nElements, size_t byteAlignment = defaultAlignment) {                                        \
+    size_t ret = 0;                                                                                                                 \
+    _ITERATE_ON_ALL(_ACCUMULATE_SOA_ELEMENT, ~, __VA_ARGS__)                                                                        \
+    return ret;                                                                                                                     \
+  }                                                                                                                                 \
+                                                                                                                                    \
+  /**                                                                                                                               \
+   * Helper/friend class allowing SoA introspection.                                                                                \
+   */                                                                                                                               \
+  struct SoAMetadata {                                                                                                              \
+    friend CLASS;                                                                                                                   \
+    SOA_HOST_DEVICE_INLINE size_t size() const { return parent_.nElements_; }                                                       \
+    SOA_HOST_DEVICE_INLINE size_t byteSize() const { return parent_.byteSize_; }                                                    \
+    SOA_HOST_DEVICE_INLINE size_t byteAlignment() const { return parent_.byteAlignment_; }                                          \
+    SOA_HOST_DEVICE_INLINE std::byte* data() const { return parent_.mem_; }                                                         \
+    SOA_HOST_DEVICE_INLINE std::byte* nextByte() const { return parent_.mem_ + parent_.byteSize_; }                                 \
```

Comment:
I think we will now be safe. `data()` only exists in the context of the layout (formerly store) that is typically not exposed to the user anymore. The only use case I had for `data()` so far is to provide `memcpy` & friends with the buffer address. The views, which are exposed to the user, have no such notion of `data()` as the columns can be in various locations. I'll close this here, but this might need further discussion in the follow-up  PR.